crates/op-rbuilder/src/generator.rs

```diff
@@ -438,4 +450,142 @@ mod tests {
+        type Primitives = OpPrimitives;
+
+        fn block(&self) -> &SealedBlockFor<<Self::Primitives as NodePrimitives>::Block> {
+            unimplemented!()
+        }
+
+        /// Returns the fees collected for the built block
+        fn fees(&self) -> U256 {
+            unimplemented!()
+        }
+
+        /// Returns the entire execution data for the built block, if available.
+        fn executed_block(&self) -> Option<ExecutedBlock<Self::Primitives>> {
+            None
+        }
+
+        /// Returns the EIP-7865 requests for the payload if any.
+        fn requests(&self) -> Option<Requests> {
+            unimplemented!()
+        }
+    }
+
+    #[derive(Debug, PartialEq, Clone)]
+    enum BlockEvent {
+        Started,
+        Cancelled,
+    }
+
+    impl<Pool, Client> PayloadBuilder<Pool, Client> for MockBuilder {
+        type Attributes = OpPayloadBuilderAttributes;
+        type BuiltPayload = MockPayload;
+
+        fn try_build(
+            &self,
+            args: BuildArguments<Pool, Client, Self::Attributes, Self::BuiltPayload>,
+            _best_payload: BlockCell<Self::BuiltPayload>,
+        ) -> Result<(), PayloadBuilderError> {
+            self.new_event(BlockEvent::Started);
+
+            loop {
+                if args.cancel.is_cancelled() {
+                    self.new_event(BlockEvent::Cancelled);
+                    return Ok(());
+                }
+
+                // Small sleep to prevent tight loop
+                std::thread::sleep(Duration::from_millis(10));
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn test_payload_generator() -> eyre::Result<()> {
+        let mut rng = thread_rng();
+
+        let pool = NoopTransactionPool::default();
+        let client = MockEthProvider::default();
+        let executor = TokioTaskExecutor::default();
+        let config = BasicPayloadJobGeneratorConfig::default();
+        let builder = MockBuilder::new();
+
+        let (start, count) = (1, 10);
+        let blocks = random_block_range(
+            &mut rng,
+            start..=start + count - 1,
+            BlockRangeParams {
+                tx_count: 0..2,
+                ..Default::default()
+            },
+        );
+
+        client.extend_blocks(blocks.iter().cloned().map(|b| (b.hash(), b.unseal())));
+
+        let generator = BlockPayloadJobGenerator::with_builder(
+            client.clone(),
+            pool,
+            executor,
+            config,
+            builder.clone(),
+        );
+
+        // this is not nice but necessary
+        let mut attr = OpPayloadBuilderAttributes::default();
+        attr.payload_attributes.parent = client.latest_header()?.unwrap().hash();
+
+        {
+            let job = generator.new_payload_job(attr.clone())?;
+            let _ = job.await;
+
+            // you need to give one second for the job to be dropped and cancelled the internal job
+            std::thread::sleep(Duration::from_secs(1));
```

Comment:
since you run your test in tokio as async task, please use tokio sleep when possible